### PR TITLE
JP-3225: Gaia proper motions enabled

### PR DESCRIPTION
**Pull request by s-goldman**
Resolves [JP-3225](https://jira.stsci.edu/browse/JP-3225)

This PR enables catalog positions to corrected for GAIA proper motions. The code follows the logic from [romancal](https://github.com/spacetelescope/romancal/blob/main/romancal/tweakreg/astrometric_utils.py) using an new epoch variable. The default for this variable is set to 2016.0 or no correction. 

**Checklist for maintainers**
- [x] added entry in `CHANGES.rst` within the relevant release section
- [x] updated or added relevant tests
- [x] updated relevant documentation
- [x] added relevant milestone
- [x] added relevant label(s)
- [x] ran regression tests, post a link to the Jenkins job below.
      [How to run regression tests on a PR](https://github.com/spacetelescope/jwst/wiki/Running-Regression-Tests-Against-PR-Branches)
- [ ] Make sure the JIRA ticket is [resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

### CODE REVIEW

File: CHANGES.rst [hunk 1]
```diff
@@ -176,6 +176,8 @@ tweakreg
 - Added the 'GAIADR3' catalog to the available options for alignment; 
   this has been enabled as the default option [#7611].
 
+- GAIADR3 catalog positions are corrrected for proper motions given an epoch [#7614].
```
Comment (hbushouse) [inline CHANGES.rst:179]: Since this PR did not make it into the 1.11.0 release, this entry will need to get moved up to the top (unreleased) section of the log.

File: docs/jwst/tweakreg/README.rst [hunk 2]
```diff
@@ -100,8 +100,9 @@ If the step parameter ``abs_refcat`` is set to 'GAIADR3', 'GAIADR2' or 'GAIADR1'
 an astrometric reference catalog then gets generated by querying
 a GAIA-based astrometric catalog web service for all astrometrically
 measured sources in the combined field-of-view of the set of input
-images. This catalog gets generated from the catalogs available
-through the GSSS STScI web services at `GSSSCATALOGS`_.
+images. This catalog is generated from the catalogs available
+through the GSSS STScI web services at `GSSSCATALOGS`_ and accounts for proper 
+motion given an epoch; defaults to 2016.0, or no proper motion correction.
```
Comment (hbushouse) [inline docs/jwst/tweakreg/README.rst:105]:
```suggestion
motion to a given epoch. The default epoch is 2016.0, which results in no proper motion correction.
```

File: jwst/tweakreg/astrometric_utils.py [hunk 3]
```diff
@@ -56,6 +56,11 @@ def create_astrometric_catalog(input_models, catalog="GAIADR3", output="ref_cat.
         Maximum number of brightest/faintest sources to return in catalog.
         If `num_sources` is negative, return that number of the faintest
         sources.  By default, all sources are returned.
+    
+    epoch: float or str, optional
```
Comment (hbushouse) [inline jwst/tweakreg/astrometric_utils.py:60]:
```suggestion
    epoch : float or str, optional
```

File: jwst/tweakreg/astrometric_utils.py [hunk 4]
```diff
@@ -81,8 +86,15 @@ def create_astrometric_catalog(input_models, catalog="GAIADR3", output="ref_cat.
     radius, fiducial = compute_radius(outwcs)
 
     # perform query for this field-of-view
-    ref_dict = get_catalog(fiducial[0], fiducial[1], sr=radius, catalog=catalog)
-    colnames = ('ra', 'dec', 'mag', 'objID')
+    epoch = (
+        epoch
+        if epoch is not None
+        else input_models[0].meta.target["proper_motion_epoch"]
```
Comment (hbushouse) [inline jwst/tweakreg/astrometric_utils.py:92]: OK, sometimes I'm a little thick, but I'm not sure I understand why you would want to use the epoch that was the reference for the target coords given in the proposal, rather than using the actual time/epoch of the observation itself, e.g. meta.observation.date_beg (which unfortunately is a date-time string, so you'd have to parse it to convert to floating-point year).

File: docs/jwst/tweakreg/README.rst (hunk 2 above)
Comment (hbushouse) [reply docs/jwst/tweakreg/README.rst:105]: Hmmm, OK, now that I've read through the rest of the code updates (below) I see that epoch is not a user-level param, but rather gets computed from the meta data of the image being processed. So this description in the docs should be expanded to indicate that's what happens, i.e. the reference epoch is automatically derived from meta data of the image and hence the user doesn't need to know about the default 2016 value.

File: jwst/tweakreg/astrometric_utils.py (hunk 4 above)
Comment (s-goldman) [reply jwst/tweakreg/astrometric_utils.py:92]: I wondered the same thing but decided to follow the logic used by the roman folks. @mairanteodoro may have a better explanation for this choice.

Comment (s-goldman) [reply jwst/tweakreg/astrometric_utils.py:92]: Sorry, I put too much faith in the similarity of the codebases and I got lazy in my testing; won't happen again. Roman gathers the epoch metadata in a different way than JWST. I'll add code to get the correct metadata.

**Comment by s-goldman** [comment]
> Resolves [JP-2916](https://jira.stsci.edu/browse/JP-2916)
> 
> This PR enables catalog positions to corrected for GAIA proper motions. The code follows the logic from [romancal](https://github.com/spacetelescope/romancal/blob/main/romancal/tweakreg/astrometric_utils.py) using an new epoch variable. The default for this variable is set to 2016.0 or no correction.
> 
> **Checklist for maintainers**
> 
> * [ ]  added entry in `CHANGES.rst` within the relevant release section
> * [ ]  updated or added relevant tests
> * [ ]  updated relevant documentation
> * [ ]  added relevant milestone
> * [x]  added relevant label(s)
> * [ ]  ran regression tests, post a link to the Jenkins job below.
>   [How to run regression tests on a PR](https://github.com/spacetelescope/jwst/wiki/Running-Regression-Tests-Against-PR-Branches)
> * [ ]  Make sure the JIRA ticket is [resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

A logistics question: Am I suppose to check these off, or is this checklist for the maintainers?

**Comment by nden** [comment]
It's for maintainers in general. However, it's also a reminder for those submitting PRs about what is expected. I think it's best if you check off items as you work through the PR and comments. It helps greatly the maintainer if those submitting PRs check these off and run regression tests with their PR.

**Comment by s-goldman** [comment]
https://plwishmaster.stsci.edu:8081/blue/organizations/jenkins/RT%2FJWST-Developers-Pull-Requests/detail/JWST-Developers-Pull-Requests/760/pipeline

File: docs/jwst/tweakreg/README.rst [hunk 5]
```diff
@@ -100,8 +100,9 @@ If the step parameter ``abs_refcat`` is set to 'GAIADR3', 'GAIADR2' or 'GAIADR1'
 an astrometric reference catalog then gets generated by querying
 a GAIA-based astrometric catalog web service for all astrometrically
 measured sources in the combined field-of-view of the set of input
-images. This catalog gets generated from the catalogs available
-through the GSSS STScI web services at `GSSSCATALOGS`_.
+images. This catalog is generated from the catalogs available
+through the GSSS STScI web services at `GSSSCATALOGS`_ and accounts for proper 
+motion to a given epoch. The default epoch is 2016.0, which results in no proper motion correction.
```
Comment (hbushouse) [inline docs/jwst/tweakreg/README.rst:105]: Given that the epoch argument is not exposed at the user level, and hence is not settable by the user (at least if they're just running the top-level tweakreg step), I would rephrase this second sentence to:
"The epoch is automatically computed from observation date and time of the input data. If for any reason the epoch cannot be computed from the meta data, it defaults to 2016.0, which results in no proper motion correction."

Comment (hbushouse) [inline docs/jwst/tweakreg/README.rst:105]:
```suggestion
through the GSSS STScI web services at `GSSSCATALOGS`_ and has the ability to account
for proper motion to a given epoch. The epoch is computed from the observation date and time
of the input data. If for any reason the epoch cannot be computed from the data, it defaults to 2016.0,
which results in no proper motion correction.
```

**Comment by hbushouse** [comment]
Another regtest run started at https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/769

File: jwst/tweakreg/astrometric_utils.py [hunk 6]
```diff
@@ -81,8 +86,15 @@ def create_astrometric_catalog(input_models, catalog="GAIADR3", output="ref_cat.
...
     # perform query for this field-of-view
-    ref_dict = get_catalog(fiducial[0], fiducial[1], sr=radius, catalog=catalog)
-    colnames = ('ra', 'dec', 'mag', 'objID')
+    epoch = (
+        epoch
+        if epoch is not None
+        else input_models[0].meta.observation.date+input_models[0].meta.observation.time
+    )
+    # keep only decimal point and digit characters
+    epoch = float("".join(c for c in str(epoch) if c == "." or c.isdigit()))
```
Comment (hbushouse) [inline jwst/tweakreg/astrometric_utils.py:95]: JWST "DATE-OBS" has the form "2023-06-29" and "TIME-OBS" has the syntax "01:23:45.678". So with these operations `epoch` becomes a combined string like "2023-06-2901:23:45.678". Line 95 then turns that into a float value of 20230629012345.678, which is no where close to what we need. We need a floating-point year, something like 2023.4958...  (i.e. almost half way through 2023). So not only is the epoch value incorrect, but I'm guessing it's also probably causing the errors seen in the regression tests: https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/769/testReport/jwst.regtest/test_nircam_align_to_gaia/

You need to call some function (I'll be there's something appropriate in astropy) to convert a properly formatted date+time string into a float year, where a properly formatted date+time string has a "T" separator between the date and time portions of the string, i.e. YYYY-MM-DDThh:mm:ss.ss

**Review by hbushouse — CHANGES_REQUESTED** [review]

File: jwst/tweakreg/astrometric_utils.py [hunk 7]
```diff
@@ -81,8 +87,15 @@ def create_astrometric_catalog(input_models, catalog="GAIADR3", output="ref_cat.
     radius, fiducial = compute_radius(outwcs)
 
     # perform query for this field-of-view
-    ref_dict = get_catalog(fiducial[0], fiducial[1], sr=radius, catalog=catalog)
-    colnames = ('ra', 'dec', 'mag', 'objID')
+    epoch = (
+        epoch
+        if epoch is not None
+        else Time(input_models[0].meta.observation.date).decimalyear
```
Comment (hbushouse) [inline jwst/tweakreg/astrometric_utils.py:93]: Looks good!  Results in a nice decimal value.

File: jwst/tweakreg/astrometric_utils.py [hunk 8]
```diff
@@ -81,8 +87,15 @@ def create_astrometric_catalog(input_models, catalog="GAIADR3", output="ref_cat.
...
+    )
+    # keep only decimal point and digit characters
+    epoch = float("".join(c for c in str(epoch) if c == "." or c.isdigit()))
```
Comment (hbushouse) [inline jwst/tweakreg/astrometric_utils.py:96]: Not sure this line is needed anymore, since the result of line 93 is already a simple decimal number (e.g. 2023.476), but it certainly can't hurt to keep it there even if it does nothing.

Comment (s-goldman) [reply jwst/tweakreg/astrometric_utils.py]: Good call. I got rid of the line and forced epoch to be an optional float, as opposed to an optional float or string.

**Comment by hbushouse** [comment]
Another regtest run started at https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/771

**Comment by s-goldman** [comment]
> https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/771

Looks like 3 regression tests failed related to small difference (~1E-5) in generated results and the truth files. I can try to take to investigate this, but it may take me a little longer to get up to speed. Once the difference are confirmed to be benign, I can update the truth files.

**Comment by hbushouse** [comment]
> > https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/771
> 
> Looks like 3 regression tests failed related to small difference (~1E-5) in generated results and the truth files. I can try to investigate this, but it may take me a little longer to get up to speed. Once the difference are confirmed to be benign, I can update the truth files.

I think the small differences showing up are likely expected due to the inclusion of proper motion. I think we can merge the PR and do a detailed investigation of the new results once it's available in jwst/master.

**Review by hbushouse — APPROVED** [review]
